packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -1547,6 +1547,35 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
     }
   }
 
+  /// Returns the smallest [Rect], in the local coordinate system, that covers
+  /// the text within the [TextRange] specified.
+  ///
+  /// This method is used to calculate the approximate position of the IME bar
+  /// on iOS.
+  ///
+  /// Returns null if [TextRange.isValid] is false for the given `range`, or the
+  /// given `range` is collapsed.
+  Rect getRectForComposingRange(TextRange range) {
+    assert(constraints != null);
+    if (!range.isValid || range.isCollapsed)
+      return null;
+    _layoutText(minWidth: constraints.minWidth, maxWidth: constraints.maxWidth);
+
+    final List<ui.TextBox> boxes = _textPainter.getBoxesForSelection(
+      TextSelection(baseOffset: range.start, extentOffset: range.end),
+    );
+
+    final ui.TextBox firstBox = boxes.first;
```

Comment:
For multiline text, first and last may not get you the bounding boxes you expect for the left and right positions

Reply:
We should add a test for this case as well.

Reply:
Could you elaborate a bit on this? I would expect the marked text to be continuous and has a consistent writing direction, because it's produced by the same IME. Is it because the text within the given `TextRange` could be using different fonts so they might be grouped in different `TextBox`es even if they are on the same line?

Reply:
I guess the line can have more than 1 runs.

Reply:
updated